File: app/src/main/java/com/ash/studios/musify/Activities/Categories/GenreList.java
package com.ash.studios.musify.Activities.Categories;

import android.content.Context;
import android.content.Intent;
import android.graphics.Color;
import android.media.MediaPlayer;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.view.View;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.cardview.widget.CardView;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.ash.studios.musify.Activities.Player;
import com.ash.studios.musify.Activities.SearchList.CategorySearch;
import com.ash.studios.musify.Adapters.GenreAdapter;
import com.ash.studios.musify.Interfaces.IControl;
import com.ash.studios.musify.Model.Genre;
import com.ash.studios.musify.Model.Song;
import com.ash.studios.musify.R;
import com.ash.studios.musify.Utils.Engine;
import com.ash.studios.musify.Utils.Instance;
import com.ash.studios.musify.Utils.Utils;
import com.bumptech.glide.Glide;
import com.google.android.material.appbar.AppBarLayout;

import java.util.ArrayList;
import java.util.Random;

import me.everything.android.ui.overscroll.OverScrollDecoratorHelper;

import static com.ash.studios.musify.Utils.Instance.songs;
import static com.ash.studios.musify.Utils.Utils.setUpUI;

public class GenreList extends AppCompatActivity implements MediaPlayer.OnCompletionListener, IControl {
    ImageView icon, shufflePlay, sequencePlay, searchBtn, optionsBtn, snippetArt, snippetPlayBtn;
    TextView title, NF, snippetTitle, snippetArtist;
    ConstraintLayout backToLib;
    AppBarLayout appBar;
    ProgressBar loader;
    CardView snippet;
    RecyclerView rv;

    Engine engine;
    Context context;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.songs_list);
        setUpUI(this);

        setIDs();
        backToLib.setOnClickListener(v -> finish());
        snippet.setOnClickListener(v -> startActivity(new Intent(context, Player.class)));
        searchBtn.setOnClickListener(v -> {
            if (rv.getAdapter() != null && rv.getAdapter().getItemCount() > 0)
                startActivity(new Intent(context, CategorySearch.class)
                        .putExtra("cat_key", 4).putExtra("cat_name", "Genres"));
        });
        new Thread() {
            @Override
            public void run() {
                super.run();
                if (rv.getAdapter() != null && rv.getAdapter().getItemCount() > 0)

                    shufflePlay.setOnClickListener(v -> {
                        Instance.shuffle = true;
                        Utils.putShflStatus(context, true);

                        new Thread(() -> {
                            ArrayList<Song> list = new ArrayList<>();
                            for (Genre genre : Utils.genres)
                                list.addAll(Utils.getGenreSongs(context, genre.getGenre_id()));

                            shufflePlay.post(() -> {
                                Instance.songs = list;
                                Instance.position = new Random().nextInt((songs.size() - 1) + 1);

                                if (Instance.songs.size() > 0) engine.startPlayer();
                                Instance.mp.setOnCompletionListener(GenreList.this);

                                updateSnippet();
                                Toast.makeText(context, "Shuffle all songs", Toast.LENGTH_SHORT).show();
                            });
                        }).start();
                    });
            }
        }.start();
        new Thread() {
            @Override
            public void run() {
                super.run();
                if (rv.getAdapter() != null && rv.getAdapter().getItemCount() > 0)

                    sequencePlay.setOnClickListener(v -> {
                        Instance.shuffle = false;
                        Utils.putShflStatus(context, false);

                        new Thread(() -> {
                            ArrayList<Song> list = new ArrayList<>();
                            for (Genre genre : Utils.genres)
                                list.addAll(Utils.getGenreSongs(context, genre.getGenre_id()));

                            sequencePlay.post(() -> {
                                Instance.songs = list;
                                Instance.position = 0;

                                if (Instance.songs.size() > 0) engine.startPlayer();
                                Instance.mp.setOnCompletionListener(GenreList.this);

                                updateSnippet();
                                Toast.makeText(context, "Sequence all songs", Toast.LENGTH_SHORT).show();
                            });
                        }).start();
                    });
            }
        }.start();
        new Thread() {
            @Override
            public void run() {
                super.run();
                snippetPlayBtn.setOnClickListener(v -> {
                    if (Instance.mp != null) {
                        if (Instance.mp.isPlaying()) {
                            snippetPlayBtn.setImageResource(R.drawable.ic_play_small);
                            Instance.mp.pause();
                        } else {
                            snippetPlayBtn.setImageResource(R.drawable.ic_pause);
                            Instance.mp.start();
                        }
                    } else {
                        engine.startPlayer();
                        snippetPlayBtn.setImageResource(R.drawable.ic_pause);
                    }
                });
            }
        }.start();
    }

    private void setIDs() {
        context = this;
        engine = new Engine(context);
        rv = findViewById(R.id.song_list);
        appBar = findViewById(R.id.app_bar);
        snippet = findViewById(R.id.snippet);
        NF = findViewById(R.id.nothing_found);
        icon = findViewById(R.id.activity_icon);
        backToLib = findViewById(R.id.lib_back);
        loader = findViewById(R.id.list_loader);
        title = findViewById(R.id.activity_title);
        searchBtn = findViewById(R.id.search_btn);
        optionsBtn = findViewById(R.id.options_btn);
        sequencePlay = findViewById(R.id.play_all_btn);
        shufflePlay = findViewById(R.id.shuffle_all_btn);
        snippetTitle = snippet.findViewById(R.id.snip_title);
        snippetArt = snippet.findViewById(R.id.snip_album_art);
        snippetArtist = snippet.findViewById(R.id.snip_artist);
        snippetPlayBtn = snippet.findViewById(R.id.snip_play_btn);

        snippetTitle.setSelected(true);
        title.setText(new StringBuilder("Genres"));
        icon.setImageDrawable(ContextCompat.getDrawable(context, R.drawable.ic_genres));
        icon.setColorFilter(Color.parseColor(getIntent().getStringExtra("icon_color")));

        if (Instance.songs != null) updateSnippet();
        rv.setLayoutManager(new LinearLayoutManager(context));

        if (Utils.genres == null || Utils.genres.size() == 0)
            new Handler(Looper.getMainLooper()).postDelayed(() ->
                    rv.setAdapter(new GenreAdapter(context, Utils.getGenres(context), loader, NF)), 10);
        else rv.setAdapter(new GenreAdapter(context, Utils.genres, loader, NF));
        if (rv.getAdapter() == null || rv.getAdapter().getItemCount() == 0) hideAttributes();
        OverScrollDecoratorHelper.setUpOverScroll(rv, OverScrollDecoratorHelper.ORIENTATION_VERTICAL);
    }

    private void hideAttributes() {
        shufflePlay.setAlpha(0.4f);
        sequencePlay.setAlpha(0.4f);
        searchBtn.setAlpha(0.4f);

        shufflePlay.setOnClickListener(null);
        sequencePlay.setOnClickListener(null);
        searchBtn.setOnClickListener(null);
    }

    private void updateSnippet() {
        if (Instance.songs.size() > 0) {
            snippet.setVisibility(View.VISIBLE);
            snippetTitle.setText(Instance.songs.get(Instance.position).getTitle());
            snippetArtist.setText(Instance.songs.get(Instance.position).getArtist());
            Glide.with(getApplicationContext())
                    .asBitmap()
                    .placeholder(R.mipmap.icon)
                    .load(Utils.getAlbumArt(Instance.songs.get(Instance.position).getAlbum_id()))
                    .into(snippetArt);

            if (Instance.mp != null && Instance.mp.isPlaying())
                snippetPlayBtn.setImageResource(R.drawable.ic_pause);
            else
                snippetPlayBtn.setImageResource(R.drawable.ic_play_small);
        } else snippet.setVisibility(View.GONE);
    }

    @Override
    public void onStartPlayer() {
        engine.startPlayer();
        updateSnippet();
    }

    @Override
    protected void onResume() {
        super.onResume();
        if (Instance.songs != null) updateSnippet();
        if (Instance.mp != null) Instance.mp.setOnCompletionListener(this);
        if (rv.getAdapter() != null) {
            rv.getAdapter().notifyDataSetChanged();
            if (rv.getAdapter().getItemCount() == 0) hideAttributes();
        }
    }

    @Override
    public void onCompletion(MediaPlayer mediaPlayer) {
        engine.playNextSong();
        if (Instance.mp != null) {
            Instance.mp = MediaPlayer.create(getApplicationContext(), Instance.uri);
            Instance.mp.start();
            Instance.mp.setOnCompletionListener(this);
        }
        Utils.putCurrentPosition(context, Instance.position);
        updateSnippet();
    }
}

File: app/src/main/java/com/ash/studios/musify/activities/categories/GenreList.kt
package com.ash.studios.musify.activities.categories

import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.graphics.PorterDuff
import android.media.MediaPlayer
import android.media.MediaPlayer.OnCompletionListener
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.View
import android.view.WindowManager
import android.widget.ImageView
import android.widget.ProgressBar
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.cardview.widget.CardView
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.ash.studios.musify.adapters.GenreAdapter
import com.ash.studios.musify.bottomSheets.GenresSort
import com.ash.studios.musify.interfaces.IControl
import com.ash.studios.musify.interfaces.IService
import com.ash.studios.musify.models.Song
import com.ash.studios.musify.R
import com.ash.studios.musify.Services.MusicService
import com.ash.studios.musify.activities.Player
import com.ash.studios.musify.activities.searchList.CategorySearch
import com.ash.studios.musify.utils.App
import com.ash.studios.musify.utils.Constants
import com.ash.studios.musify.utils.Engine
import com.ash.studios.musify.utils.Instance.mp
import com.ash.studios.musify.utils.Instance.playing
import com.ash.studios.musify.utils.Instance.position
import com.ash.studios.musify.utils.Instance.shuffle
import com.ash.studios.musify.utils.Instance.songs
import com.ash.studios.musify.utils.Instance.uri
import com.ash.studios.musify.utils.Utils
import com.bumptech.glide.Glide
import com.futuremind.recyclerviewfastscroll.FastScroller
import com.google.android.material.appbar.AppBarLayout
import me.everything.android.ui.overscroll.OverScrollDecoratorHelper
import java.util.*

class GenreList : AppCompatActivity(), OnCompletionListener, IControl, IService {
    private lateinit var icon: ImageView
    private lateinit var shufflePlay: ImageView
    private lateinit var sequencePlay: ImageView
    private lateinit var searchBtn: ImageView
    private lateinit var optionsBtn: ImageView
    private lateinit var snippetArt: ImageView
    private lateinit var snippetPlayBtn: ImageView
    private lateinit var title: TextView
    private lateinit var notFound: TextView
    private lateinit var snippetTitle: TextView
    private lateinit var snippetArtist: TextView
    private lateinit var backToLib: ConstraintLayout
    private lateinit var appBar: AppBarLayout
    private lateinit var loader: ProgressBar
    private lateinit var snippet: CardView
    private lateinit var rv: RecyclerView
    private lateinit var fs: FastScroller
    private lateinit var engine: Engine
    private lateinit var context: Context
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.songs_list)
        window.statusBarColor = Color.TRANSPARENT
        window.decorView.systemUiVisibility = View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
        window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS)
        setIDs()
        backToLib.setOnClickListener { finish() }
        snippet.setOnClickListener { startActivity(Intent(context, Player::class.java)) }
        searchBtn.setOnClickListener {
            if (rv.adapter != null && rv.adapter!!.itemCount > 0) startActivity(
                Intent(context, CategorySearch::class.java)
                    .putExtra("cat_key", 4).putExtra("cat_name", "Genres")
            )
        }
        Thread {
            if (rv.adapter != null && rv.adapter!!.itemCount > 0) shufflePlay.setOnClickListener {
                Thread {
                    val list = ArrayList<Song>()
                    for (genre in Utils.genres) list.addAll(Utils.getGenreSongs(context, genre.genre_id))
                    shufflePlay.post {
                        if (list.size > 0) {
                            songs = list
                            shuffle = true
                            position = Random().nextInt(songs!!.size - 1 + 1)
                            engine.startPlayer()
                            mp!!.setOnCompletionListener(this@GenreList)
                            updateSnippet()
                            Utils.putShflStatus(context, true)
                            Toast.makeText(context, "Shuffle all songs", Toast.LENGTH_SHORT).show()
                        } else Toast.makeText(context, "No songs found in any genres :(", Toast.LENGTH_SHORT).show()
                    }
                }.start()
            }
        }.start()
        Thread {
            if (rv.adapter != null && rv.adapter!!.itemCount > 0) sequencePlay.setOnClickListener {
                Thread {
                    val list = ArrayList<Song>()
                    for (genre in Utils.genres) list.addAll(Utils.getGenreSongs(context, genre.genre_id))
                    sequencePlay.post {
                        if (list.size > 0) {
                            songs = list
                            shuffle = false
                            position = 0
                            engine.startPlayer()
                            mp!!.setOnCompletionListener(this@GenreList)
                            updateSnippet()
                            Utils.putShflStatus(context, false)
                            Toast.makeText(context, "Sequence all songs", Toast.LENGTH_SHORT).show()
                        } else Toast.makeText(context, "No songs found in any genres :(", Toast.LENGTH_SHORT).show()
                    }
                }.start()
            }
        }.start()
        Thread {
            snippetPlayBtn.setOnClickListener {
                if (mp != null) {
                    if (mp!!.isPlaying) {
                        snippetPlayBtn.setImageResource(R.drawable.ic_play)
                        mp!!.pause()
                        stopService(Intent(context, MusicService::class.java))
                    } else {
                        snippetPlayBtn.setImageResource(R.drawable.ic_pause)
                        mp!!.start()
                        playing = true
                        startService(Intent(context, MusicService::class.java).setAction(Constants.ACTION.CREATE.label))
                    }
                } else {
                    engine.startPlayer()
                    snippetPlayBtn.setImageResource(R.drawable.ic_pause)
                }
            }
        }.start()
        optionsBtn.setOnClickListener {
            val dialog = Utils.getDialog(context, R.layout.options_dg)
            val dialogName = dialog.findViewById<TextView>(R.id.dialog_name)
            val dialogIcon = dialog.findViewById<ImageView>(R.id.dialog_icon)
            val rescanMedia = dialog.findViewById<ConstraintLayout>(R.id.rescan_media)
            val listingOption = dialog.findViewById<ConstraintLayout>(R.id.listing_options)
            val settings = dialog.findViewById<ConstraintLayout>(R.id.settings)
            dialogName.text = title.text
            dialogIcon.setImageDrawable(icon.drawable)
            settings.visibility = View.GONE
            rescanMedia.setOnClickListener {
                dialog.dismiss()
                rv.adapter = null
                notFound.visibility = View.GONE
                loader.visibility = View.VISIBLE
                Handler(Looper.getMainLooper()).postDelayed({
                    rv.adapter = GenreAdapter(context, Utils.getGenres(context), loader, notFound)
                    if (rv.adapter == null || rv.adapter!!.itemCount == 0) hideAttributes()
                }, 10)
            }
            listingOption.setOnClickListener {
                dialog.dismiss()
                val genresSort = GenresSort(context, rv, loader, notFound)
                genresSort.show(supportFragmentManager, null)
            }
        }
    }

    private fun setIDs() {
        context = this
        engine = Engine(context)
        rv = findViewById(R.id.song_list)
        appBar = findViewById(R.id.app_bar)
        snippet = findViewById(R.id.snippet)
        notFound = findViewById(R.id.nothing_found)
        fs = findViewById(R.id.fast_song_list)
        icon = findViewById(R.id.activity_icon)
        backToLib = findViewById(R.id.lib_back)
        loader = findViewById(R.id.list_loader)
        title = findViewById(R.id.activity_title)
        searchBtn = findViewById(R.id.search_btn)
        optionsBtn = findViewById(R.id.options_btn)
        sequencePlay = findViewById(R.id.play_all_btn)
        shufflePlay = findViewById(R.id.shuffle_all_btn)
        snippetTitle = snippet.findViewById(R.id.snip_title)
        snippetArt = snippet.findViewById(R.id.snip_album_art)
        snippetArtist = snippet.findViewById(R.id.snip_artist)
        snippetPlayBtn = snippet.findViewById(R.id.snip_play_btn)
        snippetTitle.isSelected = true
        title.text = StringBuilder("Genres")
        icon.setImageDrawable(ContextCompat.getDrawable(context, R.drawable.ic_genres))
        icon.setColorFilter(Color.parseColor(intent.getStringExtra("icon_color")))
        if (songs != null) updateSnippet()
        rv.layoutManager = LinearLayoutManager(context)
        if (Utils.genres == null || Utils.genres.size == 0) Handler(Looper.getMainLooper()).postDelayed(
            { rv.adapter = GenreAdapter(context, Utils.getGenres(context), loader, notFound) },
            10
        ) else rv.adapter = GenreAdapter(context, Utils.genres, loader, notFound)
        if (rv.adapter == null || rv.adapter!!.itemCount == 0) hideAttributes()
        OverScrollDecoratorHelper.setUpOverScroll(rv, OverScrollDecoratorHelper.ORIENTATION_VERTICAL)
        fs.setRecyclerView(rv)
    }

    private fun hideAttributes() {
        shufflePlay.alpha = 0.4f
        sequencePlay.alpha = 0.4f
        searchBtn.alpha = 0.4f
        shufflePlay.setOnClickListener(null)
        sequencePlay.setOnClickListener(null)
        searchBtn.setOnClickListener(null)
    }

    private fun updateSnippet() {
        if (songs!!.size > 0) {
            snippet.visibility = View.VISIBLE
            snippetTitle.text = songs!![position].title
            snippetArtist.text = songs!![position].artist
            Glide.with(applicationContext)
                .asBitmap()
                .placeholder(R.drawable.placeholder)
                .load(Utils.getAlbumArt(songs!![position].album_id))
                .into(snippetArt)
            val accents = Utils.getSecondaryColors(context, Utils.getAlbumArt(songs!![position].album_id))
            snippet.setCardBackgroundColor(accents[0])
            snippetTitle.setTextColor(accents[1])
            snippetArtist.setTextColor(accents[1])
            snippetPlayBtn.setColorFilter(accents[1], PorterDuff.Mode.SRC_IN)
            if (mp != null && mp!!.isPlaying) snippetPlayBtn.setImageResource(R.drawable.ic_pause) else snippetPlayBtn.setImageResource(R.drawable.ic_play)
        } else snippet.visibility = View.GONE
    }

    override fun onStartService() {
        engine.startPlayer()
        updateSnippet()
    }

    override fun onPrevClicked() {
        engine.playPrevSong()
        updateSnippet()
        startService(Intent(context, MusicService::class.java).setAction(Constants.ACTION.CREATE.label))
    }

    override fun onNextClicked() {
        engine.playNextSong()
        updateSnippet()
        startService(Intent(context, MusicService::class.java).setAction(Constants.ACTION.CREATE.label))
    }

    override fun onPlayClicked() {
        if (mp != null) mp!!.start() else {
            engine.startPlayer()
            mp!!.setOnCompletionListener(this@GenreList)
        }
        updateSnippet()
    }

    override fun onPauseClicked() {
        if (mp != null) mp!!.pause()
        updateSnippet()
    }

    override fun onStopService() {
        if (mp != null) {
            mp!!.stop()
            mp!!.release()
            mp = null
        }
        snippetPlayBtn.setImageResource(R.drawable.ic_play)
        stopService(Intent(context, MusicService::class.java))
    }

    @SuppressLint("NotifyDataSetChanged")
    override fun onResume() {
        super.onResume()
        (applicationContext as App).currentContext = context
        if (songs != null) updateSnippet()
        if (mp != null) mp!!.setOnCompletionListener(this)
        if (rv.adapter != null) {
            rv.adapter!!.notifyDataSetChanged()
            if (rv.adapter!!.itemCount == 0) hideAttributes()
        }
    }

    override fun onCompletion(mediaPlayer: MediaPlayer) {
        engine.playNextSong()
        if (mp != null) {
            mp = MediaPlayer.create(applicationContext, uri)
            mp!!.start()
            mp!!.setOnCompletionListener(this)
        }
        Utils.putCurrentPosition(context, position)
        updateSnippet()
    }
}